### PR TITLE
Spatial Index H3 example

### PR DESCRIPTION
**Pull request by menusal**
## Description

We will use the following h3 dataset at resolucion 8: cartobq.public_account.derived_spatialfeatures_usa_h3int_res8_v1_yearly_v2

It's a dataset of 12M rows and 13.1GB of data.

We will show how we can run dynamic queries in real-time with a query as follows:

```sql
select h3, {variable} cartobq.public_account.derived_spatialfeatures_usa_h3int_res8_v1_yearly_v2
where {variable} between {@min_value} and {@max_value}
```

The list of variables is available at:
`https://public.carto.com/api/v4/data/observatory/metadata/datasets/cdb_spatial_fea_94e6b1f/variables?minimal=true`. We will only display float variables and we will show the name and the description of the variable in the UI.

All the filtering should run server side

Interactivity as hover is requrired to show: hexid + variable

Shorcut: [sc-361044](https://app.shortcut.com/cartoteam/story/361044/spatial-index-h3-example)

## CHECKLIST

- [x] Spatial Index H3 Layer
- [x] Filtering

## TODO

- Discussion about filter

### CODE REVIEW

**Comment by menusal** [comment]
## Discussion about filtering

The filter at the highest aggregation number (maximum zoom) is well understood.
The problem conceptually comes at the aggregation level. When we zoom out, since there is an aggregation expression `SUM(<field>)`, the filters are acting at the single cell level and do not take into account the aggregation `aggregationExp` and paint cells, which visually the user would expect not to see since they are adding the values that contain the filter.

image

I'd like you to please watch the video attached in the description.

File: spatial-index-h3/index.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
...rdfF9HNCJtf2XpGP/B5NuZ8XO/hUBX0h6nmkE/geApPhrHBiMqdCIVFKvySHrzw4oAErZMfkR52dEmWywkbmnCCntFPa4hskLXlQyec8rRLTztGBQkXRgwXRrxbf76HU5VvxS0mwql6IxFP6+Gw6Lguia1ho2Fdl3T1p/W0aTwwwxH2BKHACZCkZ9hmv4flfL13EzaHmA7hVX5ef7+HW3u2RPzh+sYCtMTsxdd7uMinC9tMHKuMy233b9K3Cr/aD5yoRwSPezjW+DS9Tq4uPd4yjHX0BCjxZXjq4VjtUfpY/kJmYoY7kuFCJa7kqo7tYz5GPvRQ1di9m/MfnWPbA1mYdo1a9ivx9R6Oh1VUPFW+CokdxzJ7RnrYfpLBj2rsUZO15UHeKC9ZmHa55Pzw6Ivw+R5OVbYVu5iUYJY5rD1+X+FjOGCTMTmNDzQaiLlhm0w9nOk26mEK9lvY+PlahO/3cBGuwjZ/rvdwWCQb2hOvym5NO6laKgL3pEdWhNce7cad5cICr/dwnhCl8VY+0GMHQ8VqE2tdhKcejocN9jzLn1cczpOwhxrGRVoqT3Ij+nP8ZOaCw43oZoTNr2w9ox/4PBtzPq73cKgC+sNUc8gncDyFJ8PYYETlSoTCIhV+SQ9eeHFAgtbJj0gPe+ndLKe8bp/ZekY/rSGy4ezKi/7VsuFNevIYFSRcGDEoRrxbf76HU5VvxS0mwql6IxFPb2ZsL+A6Fj9Rjx2HCg8ajuAThCJRz7DNfgsTO4t4leYQ07FeF/68/n4Pt/ZsifjD9Y0FaInZi6/3cJFPF7aJOFYZl9u3Y5O+VfjVfuBEPSK43sNpHS0ZnhAtRBqeejh88dsfqLMxw2OtDTKRVsmT3FUx3udh5DhBOZaezRY+W+9edf5KzOiMOqJ0WYSv93AscvZRLbnRSWqzeunhslshfzVG+GJX/FV47NGQ+kp8vodTlW3FLiYlmGUNa4/fV/jwU70Ck8P4uNMBbCDmhm2C3DS8Vw8q79ncbv9/KPPx/R4uwlXY9nP6eg+HRbKhPfCq7Na006qlInBPemRFeO3RbtxZLizweg/nCVEab+UDPXYwVCyysdZFeOrheNhgz7P8ecXhPAl7qGFcpKXyJDeiP8dP5vbr4j96Nrtvut/QhJ/QM3rxRx7UUykv/8f3cP8C5bGlenBMtnYAAAAASUVORK5CYII="
```
Comment (alasarr) [inline spatial-index-h3/index.html:39]: Let's add a separate folder for the image

File: spatial-index-h3/.env [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# API Base URL (copy this from CARTO Workspace -> Developers)
```
Comment (alasarr) [inline spatial-index-h3/.env:1]: The name of the folder should be: spatial-features-h3

File: spatial-index-h3/index.html [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  <head>
+    <meta charset="UTF-8" />
+    <title>CARTO for deck.gl - H3 Spatial Index 12M rows</title>
+  </head>
+  <body>
+    <div id="map"></div>
+    <canvas id="deck-canvas"></canvas>
+    <div id="story-card">
+      <p class="overline">✨👀 You're viewing</p>
+      <h2>CARTO for deck.gl - H3 Spatial Index 12M rows</h2>
```
Comment (alasarr) [inline spatial-index-h3/index.html:12]: CARTO Spatial Features datasets in H3 (12M rows).

File: spatial-index-h3/index.html [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      <p>
+        More info about
+        <a
+          href="https://docs.carto.com/carto-user-manual/workflows/components/spatial-indexes"
+          rel="noopener noreferrer"
+          target="_blank"
+          >H3 Spatial Index</a
```
Comment (alasarr) [inline spatial-index-h3/index.html:19]: Link to https://carto.com/blog/carto-spatial-features-urbanity-climatology-elevation

File: spatial-index-h3/index.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  apiBaseUrl: import.meta.env.VITE_API_BASE_URL,
+  accessToken: import.meta.env.VITE_API_ACCESS_TOKEN,
+  connectionName: import.meta.env.VITE_API_CONNECTION_NAME
+};
+
+function render() {
+  const source = h3QuerySource({
+    ...cartoConfig,
+    aggregationExp: `SUM(${selectedVariable}) as value`,
+    spatialDataColumn: 'h3:h3',
```
Comment (alasarr) [inline spatial-index-h3/index.ts:53]: h3:h3 shouldn't be supported. Is it not complaining?

File: spatial-index-h3/index.html [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        >
+      </p>
+
+      <div class="layer-controls">
+        <p class="overline">Variable</p>
+        <select name="variable" id="variable" class="select">
+          <option value="population" selected>Population</option>
```
Comment (alasarr) [inline spatial-index-h3/index.html:26]: It should be populated form the API

File: spatial-index-h3/index.ts (hunk 5 above)
Comment (alasarr) [reply spatial-index-h3/index.ts:53]: Yes it's:
image

**Review by alasarr — CHANGES_REQUESTED** [review]

File: spatial-index-h3/index.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    sqlQuery: `SELECT * FROM cartobq.public_account.derived_spatialfeatures_usa_h3int_res8_v1_yearly_v2 WHERE ${selectedVariable} between @min and @max`,
```
Comment (felixpalmer) [inline spatial-index-h3/index.ts:54]: Seems tidier to call these parameters `@from` and `@to`

File: spatial-index-h3/index.ts (hunk 5 above)
Comment (menusal) [reply spatial-index-h3/index.ts:53]: Removed

File: spatial-index-h3/index.html (hunk 6 above)
Comment (menusal) [reply spatial-index-h3/index.html:26]: Done `min` 0 max `SELECT max(population) FROM cartobq.public_account.derived_spatialfeatures_usa_h3int_res8_v1_yearly_v2` 43303

File: spatial-index-h3/index.ts (hunk 7 above)
Comment (menusal) [reply spatial-index-h3/index.ts:54]: Done

File: spatial-index-h3/.env (hunk 2 above)
Comment (menusal) [reply spatial-index-h3/.env]: Renamed

File: spatial-index-h3/index.html (hunk 6 above)
Comment (alasarr) [reply spatial-index-h3/index.html]: Let's do the following:
- Load variables only the first time
- Remove the from and to filter
- Use exponential breaks: 100, 1000, 10000.... 
- Choose a default zoom that shows enough variation in the colors of the palette.

Comment (menusal) [reply spatial-index-h3/index.html]: @alasarr About:

`Load variables only the first time`

All the fields, excluding h3, and no "Number" fields?
```js
{
    "h3": "Number",
    "do_date": "Timestamp",
    "country_iso": "String",
    "country_iso_a3": "String",
    "population": "Number",
    "female": "Number",
    "male": "Number",
    "female_under_1": "Number",
    "female_1_to_4": "Number",
    "female_5_to_9": "Number",
    "female_10_to_14": "Number",
    "female_15_to_19": "Number",
    "female_20_to_24": "Number",
    "female_25_to_29": "Number",
    "female_30_to_34": "Number",
    "female_35_to_39": "Number",
    "female_40_to_44": "Number",
    "female_45_to_49": "Number",
    "female_50_to_54": "Number",
    "female_55_to_59": "Number",
    "female_60_to_64": "Number",
    "female_65_to_69": "Number",
    "female_70_to_74": "Number",
    "female_75_to_79": "Number",
    "female_80_and_over": "Number",
    "male_under_1": "Number",
    "male_1_to_4": "Number",
    "male_5_to_9": "Number",
    "male_10_to_14": "Number",
    "male_15_to_19": "Number",
    "male_20_to_24": "Number",
    "male_25_to_29": "Number",
    "male_30_to_34": "Number",
    "male_35_to_39": "Number",
    "male_40_to_44": "Number",
    "male_45_to_49": "Number",
    "male_50_to_54": "Number",
    "male_55_to_59": "Number",
    "male_60_to_64": "Number",
    "male_65_to_69": "Number",
    "male_70_to_74": "Number",
    "male_75_to_79": "Number",
    "male_80_and_over": "Number",
    "retail": "Number",
    "education": "Number",
    "financial": "Number",
    "food_drink": "Number",
    "healthcare": "Number",
    "leisure": "Number",
    "tourism": "Number",
    "transportation": "Number",
    "elevation": "Number",
    "urbanity": "String",
    "tmin_jan": "Number",
    "tmin_feb": "Number",
    "tmin_mar": "Number",
    "tmin_apr": "Number",
    "tmin_may": "Number",
    "tmin_jun": "Number",
    "tmin_jul": "Number",
    "tmin_aug": "Number",
    "tmin_sep": "Number",
    "tmin_oct": "Number",
    "tmin_nov": "Number",
    "tmin_dec": "Number",
    "tmax_jan": "Number",
    "tmax_feb": "Number",
    "tmax_mar": "Number",
    "tmax_apr": "Number",
    "tmax_may": "Number",
    "tmax_jun": "Number",
    "tmax_jul": "Number",
    "tmax_aug": "Number",
    "tmax_sep": "Number",
    "tmax_oct": "Number",
    "tmax_nov": "Number",
    "tmax_dec": "Number",
    "tavg_jan": "Number",
    "tavg_feb": "Number",
    "tavg_mar": "Number",
    "tavg_apr": "Number",
    "tavg_may": "Number",
    "tavg_jun": "Number",
    "tavg_jul": "Number",
    "tavg_aug": "Number",
    "tavg_sep": "Number",
    "tavg_oct": "Number",
    "tavg_nov": "Number",
    "tavg_dec": "Number",
    "prec_jan": "Number",
    "prec_feb": "Number",
    "prec_mar": "Number",
    "prec_apr": "Number",
    "prec_may": "Number",
    "prec_jun": "Number",
    "prec_jul": "Number",
    "prec_aug": "Number",
    "prec_sep": "Number",
    "prec_oct": "Number",
    "prec_nov": "Number",
    "prec_dec": "Number",
    "srad_jan": "Number",
    "srad_feb": "Number",
    "srad_mar": "Number",
    "srad_apr": "Number",
    "srad_may": "Number",
    "srad_jun": "Number",
    "srad_jul": "Number",
    "srad_aug": "Number",
    "srad_sep": "Number",
    "srad_oct": "Number",
    "srad_nov": "Number",
    "srad_dec": "Number",
    "wind_jan": "Number",
    "wind_feb": "Number",
    "wind_mar": "Number",
    "wind_apr": "Number",
    "wind_may": "Number",
    "wind_jun": "Number",
    "wind_jul": "Number",
    "wind_aug": "Number",
    "wind_sep": "Number",
    "wind_oct": "Number",
    "wind_nov": "Number",
    "wind_dec": "Number",
    "vapr_jan": "Number",
    "vapr_feb": "Number",
    "vapr_mar": "Number",
    "vapr_apr": "Number",
    "vapr_may": "Number",
    "vapr_jun": "Number",
    "vapr_jul": "Number",
    "vapr_aug": "Number",
    "vapr_sep": "Number",
    "vapr_oct": "Number",
    "vapr_nov": "Number",
    "vapr_dec": "Number"
}
```

Comment (alasarr) [reply spatial-index-h3/index.html]: Yes, but use the [endpoint](https://public.carto.com/api/v4/data/observatory/metadata/datasets/cdb_spatial_fea_94e6b1f/variables?minimal=true), don't hard code it and use the name property in the UI

Comment (menusal) [reply spatial-index-h3/index.html]: @alasarr Done, but:

I think we should have several domains configured depending on the type of variable, static ranges work well with `population`, `male`, and `female` but not for the rest, we would have to generate `domains` for:

- Population
- Age cohorts
- Temperature
- Wind
- Solar radiation
- Monthly vapour pressure
- etc

File: spatial-features-h3/deck.gl__carto [hunk 8]
```diff
@@ -0,0 +1 @@
+../../../../deck.gl/modules/carto/typed
```
Comment (alasarr) [inline spatial-features-h3/deck.gl__carto:1]: Remove this

File: spatial-features-h3/index.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+const variableSelector = document.getElementById('variable') as HTMLSelectElement;
+variableSelector?.addEventListener('change', () => {
+  selectedVariable = variableSelector.value;
+  render();
+});
+
+function initSelectors() {
+  const variableSelector = document.getElementById('variable') as HTMLSelectElement;
+  fetch(
+    'https://public.carto.com/api/v4/data/observatory/metadata/datasets/cdb_spatial_fea_94e6b1f/variables?minimal=true'
```
Comment (alasarr) [inline spatial-features-h3/index.ts:35]: For better explanation, could we have a getVariables async function which takes care of the fetch?

File: spatial-features-h3/index.html [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+            src="./images/scale.jpg"
+            alt="legend"
+          />
+          <div class="label-container">
+            <div>0</div>
+            <div>4K</div>
+          </div>
+        </div>
+      </div>
+    </div>
```
Comment (alasarr) [inline spatial-features-h3/index.html:40]: Add the source link available in other examples with a link to https://carto.com/spatial-data-catalog/browser/dataset/cdb_spatial_fea_94e6b1f/

File: spatial-features-h3/index.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  )
+    .then(response => response.json())
+    .then(data => {
+      const options = data
+        .filter(variable => variable.db_type === 'FLOAT')
+        .map((variable: any) => {
```
Comment (alasarr) [inline spatial-features-h3/index.ts:41]: Can we type this? At least with the minimal columnName and variableName.

File: spatial-features-h3/index.ts (hunk 9 above)
Comment (alasarr) [reply spatial-features-h3/index.ts:35]: I would consider a separate file for this

File: spatial-features-h3/package.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (alasarr) [inline spatial-features-h3/package.json:1]: Align with other examples

File: spatial-features-h3/style.css [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+  margin-top: 2.25rem;
+  margin-bottom: 0px;
+  font-size: 0.8rem;
+}
+
+.city-button-group {
+  display: flex;
+}
+
+.population-slider {
```
Comment (alasarr) [inline spatial-features-h3/style.css:87]: I feel we don't need all the styles

**Review by alasarr — CHANGES_REQUESTED** [review]

File: spatial-features-h3/index.ts (hunk 9 above)
Comment (menusal) [reply spatial-features-h3/index.ts:35]: Done

File: spatial-features-h3/index.html (hunk 10 above)
Comment (menusal) [reply spatial-features-h3/index.html:40]: Done

File: spatial-features-h3/package.json (hunk 12 above)
Comment (menusal) [reply spatial-features-h3/package.json:1]: Done

File: spatial-features-h3/index.ts (hunk 11 above)
Comment (menusal) [reply spatial-features-h3/index.ts]: Done. Typed in utils

File: spatial-features-h3/style.css (hunk 13 above)
Comment (menusal) [reply spatial-features-h3/style.css:87]: removed

File: spatial-features-h3/deck.gl__carto (hunk 8 above)
Comment (menusal) [reply spatial-features-h3/deck.gl__carto]: done

File: spatial-index-h3/index.html (hunk 6 above)
Comment (alasarr) [reply spatial-index-h3/index.html]: That's why I suggested logarithmic breaks for `colorBins`. I added it in this commit: ad616595c9577d10710336a3a8313cc964281d3c

Comment (alasarr) [reply spatial-index-h3/index.html]: It's not perfect, but that's fine for an example

**Pull request merged**